File: examples/python/demo_buffer_decrypt.py
#!/usr/bin/env python3
"""
mp4decrypt-api - In-Memory RAM Buffer Decryption Demo (Python)
=============================================================

Demonstrates Paradigm 2: Decrypting media stored entirely in RAM memory
buffers using both the object-oriented Mp4DecryptSession and the one-shot
convenience method Mp4DecryptSession.decrypt_buffer().

Run with:
    python examples/python/demo_buffer_decrypt.py
"""

import os
import sys
import time
import urllib.request

# Ensure sdk/python is importable
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")))
from sdk.python.mp4decrypt_sdk import Mp4DecryptSession


def main():
    print("=" * 67)
    print("     MP4DECRYPT-API - IN-MEMORY BUFFER DECRYPTION (PYTHON)         ")
    print("=" * 67)

    base_url = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey"
    kid_hex = "9eb4050de44b4802932e27d75083e266"
    key_hex = "166634c675823c235a4a9446fad52e4d"

    print("1. Downloading encrypted fragments into memory...")
    init_bytes = urllib.request.urlopen(f"{base_url}/15/init.mp4").read()
    seg_bytes = urllib.request.urlopen(f"{base_url}/15/0001.m4s").read()
    encrypted_buffer = init_bytes + seg_bytes

    print(f"   Buffer size in RAM: {len(encrypted_buffer):,} bytes")

    # Method A: Object-oriented session
    print("\n2. Decrypting via Mp4DecryptSession context...")
    t0 = time.perf_counter()
    with Mp4DecryptSession() as session:
        session.add_key(kid_hex, key_hex)
        decrypted_a = session.decrypt_memory(encrypted_buffer)
        stats = session.get_stats()
    duration_a = (time.perf_counter() - t0) * 1000.0

    print(f"   Session output size: {len(decrypted_a):,} bytes")
    print(f"   Execution time:     {duration_a:.2f} ms (native pipeline: {stats.execution_duration_ms:.2f} ms)")

    # Method B: One-shot convenience method
    print("\n3. Decrypting via one-shot Mp4DecryptSession.decrypt_buffer()...")
    t1 = time.perf_counter()
    decrypted_b = Mp4DecryptSession.decrypt_buffer(encrypted_buffer, kid_hex, key_hex)
    duration_b = (time.perf_counter() - t1) * 1000.0

    print(f"   One-shot output size: {len(decrypted_b):,} bytes")
    print(f"   Total duration:       {duration_b:.2f} ms")

    # Verification
    print("\n4. Verifying media contents in-memory (zero disk I/O)...")
    info = Mp4DecryptSession.probe(decrypted_b)
    print(f"   Track count:         {info.track_count}")
    print(f"   Track #1 codec:      {info.tracks[0].codec}")
    print(f"   Track #1 encrypted:  {info.tracks[0].is_encrypted}")

    if not info.tracks[0].is_encrypted and decrypted_a == decrypted_b:
        print("\nSUCCESS: Both in-memory decryption paradigms produced identical plaintext!")
    else:
        raise RuntimeError("Decryption verification failed.")


if __name__ == "__main__":
    main()

File: examples/bun/demo_buffer_decrypt.ts
/**
 * mp4decrypt-api - One-Shot Buffer Decryption Demo (Bun)
 * =======================================================
 *
 * Demonstrates 1-line zero-setup in-memory buffer decryption and
 * verifies that decrypted media is plaintext.
 *
 * Run with:
 *   bun examples/bun/demo_buffer_decrypt.ts
 */

import { Mp4DecryptSession } from "../../sdk/bun/mp4decrypt_sdk";

async function main() {
  console.log("===================================================================");
  console.log("    MP4DECRYPT-API - ONE-SHOT BUFFER DECRYPTION DEMO (BUN)         ");
  console.log("===================================================================");

  const baseUrl = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey";
  const kidHex = "9eb4050de44b4802932e27d75083e266";
  const keyHex = "166634c675823c235a4a9446fad52e4d";

  console.log("Fetching encrypted init.mp4 and 0001.m4s into RAM...");
  const [initResp, segResp] = await Promise.all([
    fetch(`${baseUrl}/15/init.mp4`),
    fetch(`${baseUrl}/15/0001.m4s`),
  ]);

  const initBytes = new Uint8Array(await initResp.arrayBuffer());
  const segBytes = new Uint8Array(await segResp.arrayBuffer());

  // Combine init + media fragment into a single playable encrypted fMP4 buffer
  const encrypted = new Uint8Array(initBytes.byteLength + segBytes.byteLength);
  encrypted.set(initBytes, 0);
  encrypted.set(segBytes, initBytes.byteLength);

  console.log(`Encrypted buffer size: ${(encrypted.byteLength / 1024).toFixed(1)} KB`);

  // Verify before decryption
  const preProbe = Mp4DecryptSession.probe(encrypted);
  console.log(`Pre-decryption status: Track encrypted = ${preProbe.tracks[0]?.isEncrypted}, Scheme = ${preProbe.tracks[0]?.schemeType}`);

  // One-shot decryption
  console.log("\nExecuting Mp4DecryptSession.decryptBuffer(buffer, kidHex, keyHex)...");
  const t0 = performance.now();
  const decrypted = Mp4DecryptSession.decryptBuffer(encrypted, kidHex, keyHex);
  const t1 = performance.now();

  console.log(`Decrypted buffer size: ${(decrypted.byteLength / 1024).toFixed(1)} KB in ${(t1 - t0).toFixed(2)} ms`);

  // Verify after decryption
  const postProbe = Mp4DecryptSession.probe(decrypted);
  console.log(`Post-decryption status: Track encrypted = ${postProbe.tracks[0]?.isEncrypted}`);
  console.log(`Decrypted codec: ${postProbe.tracks[0]?.codec}`);

  if (postProbe.tracks[0]?.isEncrypted === false) {
    console.log("\nSUCCESS: Media was successfully decrypted in memory!");
  } else {
    throw new Error("Decryption verification failed: media still marked encrypted.");
  }
}

main().catch(console.error);
